### ISSUE TITLE
Missing address filtering for Open_T_Group

### ISSUE THREAD

**Issue by TrondKjeldas** [knxd/knxd#453]
When a client uses the Open_T_Group to subscribe to a GA, it receives telegrams for all GAs, not just the one it subscribes to.

This seems to be a regression for a previously fixed issue, described in #359, which was previously fixed in commit 358b87d.

There are quite a lot of changes to the code in between, so I have not  gotten around to debug it fully yet.

**Comment by smurfix** [comment]
Thanks for the report. I'm looking into this.

**Comment by smurfix** [comment]
@tobylorenz This worked at the beginning of your patch series, it does not work at the end, and your series is not bisect-able.
Thus, please find the problem and fix it.
Thank you.

**Comment by TrondKjeldas** [comment]
Have done a bit of debugging.

Seems like it reaches this line:

        if (l1->hop_count == 7 || ii->checkGroupAddress(l1->destination_address))

around line 1133 in router.cpp.

The ii->checkGroupAddress() resolves to the one defined around line 900 in links.h, which is defined to simply always return true.

It is not clear to me how to fix this, since the address available in the ii->checkGroupAddress() is a device address, while the parameter is a group address.

**Comment by smurfix** [comment]
It shouldn't go to LineDriver::checkGroupAddress because the class in question should be a `T_Group` (layer4.h) which is a subclass of LineDriver and overrides `checkGroupAddress`.

**Comment by TrondKjeldas** [comment]
Makes sense, since the previous fix was adding a missing address check in layer4.h

**Comment by tobylorenz** [comment]
Seems I missed here something...
However all tests were green. Do we miss a test case here as well?

**Comment by TrondKjeldas** [comment]
I found the issue. Someone has made the layer4 version of the checkGroupAddress const:

    virtual bool checkGroupAddress(eibaddr_t addr) const override

while the base class (LineDriver) version of the checkGroupAddress is not:

    virtual bool checkGroupAddress (eibaddr_t)
 
Since the `const` is part of the signature this means that the layer4 version does not override the LineDriver version.

In other words, either all, or none, of the methods must be const.

Might be worth checking if any other virtual methods has been constified as well.

**Comment by TrondKjeldas** [comment]
To expand a little.

The layer4 version I am referring to is in the T_Group class.

The reason the override specifier does not cause a warning/error, is because the T_Group class is sub-classed from Layer4commonWO, and Layer4commonWO::checkGroupAddress is also const, but does not have the override specifier.

If the Layer4commonWO::checkGroupAddress method also had the override specifier the compiler would correctly complain, since it does not override the LineDriver::checkGroupAddress method.

**Comment by TrondKjeldas** [comment]
I looked through the changes done during 2019 to check for methods that has been constified.

There was quite a lot, but they mostly looked to be consistent.

I found two that potentially have the same problem as the checkGroupAddress method:

checkAddress  in link.h and layer4.h 
getVersion in ft12.h, cemi.h, and emi*.h

Note that I'm not that familiar with the code base, hence they might be false positives.

**Comment by smurfix** [comment]
@tobylorenz I'm quite sure that the test suite is woefully incomplete.
@TrondKjeldas Thanks, could you create a pull request?

**Comment by TrondKjeldas** [comment]
Sure, will do.

**Comment by tobylorenz** [comment]
I somehow feel guilty to have introduced this regression, so I started working on a fix today as well. It adds const/virtual/override to the aforementioned methods, so it basically finalizes my work. I'll try to send a pull request soon.

**Comment by tobylorenz** [comment]
I think the patch solves the issue. At least it shouldn't make anything worse ;-)
